### ISSUE TITLE
Custom fitness function as model.train() parameter

### ISSUE THREAD

**Issue by gmarinho2** [ultralytics/ultralytics#23870]
### The problem

Currently, the selection of the `best.pt` model is heavily biased towards bounding box precision, as the default fitness weights ([ref 1](https://github.com/ultralytics/ultralytics/blob/7710ef05dc56dfe61b20a7537f17078db2b7170b/ultralytics/utils/metrics.py#L964)) corresponding to $[Precision, Recall, mAP_{50}, mAP_{50-95}]$ are:

$$w = [0.0, 0.0, 0.0, 1.0]$$ 

This 1.0 weight on mAP50-95 forces the model to give more relevance to pixel-perfect boxes, which is suboptimal for some tracking-heavy use cases, especially those involving challenging environments (e.g., maritime), the $mAP_{50-95}$ is too restrictive. In these cases, a model that achieves a higher $mAP_{50}$ is often more useful for maintaining object IDs, even if the bounding box precision is slightly lower.

### Solution

I would like to see a new argument in the `model.train()` method, such as `fitness_weights`, that allows users to pass a list or tuple to override the default metric importance. 

Example usage:
```python
# Prioritizes mAP50 for saving best.pt
model.train(data="data.yaml", fitness_weights=[0.0, 0.0, 1.0, 0.0]) 
```

### Use case

Different systems have different concepts of which model is the best. Problems that prioritize knowing where the object is can benefit from saving the best model based on having a fitness function that gives more weight to mAP50. Example: Surfer tracking (what I was doing to realise that), where the continuity of the detection is more critical than the pixel-perfect alignment of the bounding box. 

In high-dynamic environments with visual obstructions like sea spray, foam, and motion blur, a strict mAP50-95 requirement can lead to 'flickering' detections or false negatives. By anchoring the fitness function to mAP50, the training process prioritizes a model that maintains a stable lock on the subject, ensuring that downstream tracking algorithms (like BoT-SORT) receive consistent ID inputs even when the object's precise boundaries are visually ambiguous.

### Implementation Details

**1. Argument Behavior & Validation**
- The `fitness_weights` argument should accept a `List[float]` of length 4.
- **Validation:** - Check if `len(weights) == 4`.
    - Ensure all values are non-negative.
    - Default to `[0.0, 0.0, 0.0, 1.0]` if the argument is `None` to ensure backward compatibility.

**2. Checkpoint Selection (`best.pt`)**
- The custom weights should only influence the `fitness` calculation used to update `self.best_fitness`.
- The `last.pt` checkpoint remains unaffected, saving the state of the final epoch as usual.
- **Persistence:** These weights should be stored within the model's YAML/metadata. This ensures that when using `model.train(resume=True)`, the training continues with the same fitness criteria.

**3. Reporting and Logging**
- **Console Log:** At the start of training, a message should confirm: `Plotting fitness using custom weights: [P, R, mAP50, mAP50-95]`.
- **CSV Logging:** The `fitness` column in `results.csv` should reflect the value calculated with the custom weights. This allows users to visualize how their specific optimization target is evolving.

**4. Scope**
- This change applies to the `DetMetrics` class in `ultralytics/utils/metrics.py` and the trainer logic in `engine/trainer.py`.

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @gmarinho2, thank you for your interest in Ultralytics 🚀 and for the thoughtful feature request around customizable training fitness behavior. This is an automated response to help get things moving quickly, and an Ultralytics engineer will also assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

For this feature request, it would be very helpful if you could share any additional implementation details you have in mind, such as expected argument behavior, validation rules, checkpoint selection behavior, and whether you’d like this to apply only to `best.pt` selection or also to related reporting/logging. Since you mentioned you’d like to submit a PR 🙌, that extra context can help the team review the proposal faster.

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
You can check the custom trainer example on how to define a different fitness function 

https://docs.ultralytics.com/guides/custom-trainer/#saving-the-best-model-by-custom-metric
